src/serializer/serializer.js

```diff
@@ -204,14 +198,28 @@ export class Serializer {
             this.realm,
             this.logger,
             this.modules,
-            additionalFunctionValuesAndEffects,
-            referentializer
+            additionalFunctionValuesAndEffects
           );
           statistics.deepTraversal.measure(() => residualHeapVisitor.visitRoots());
           return [residualHeapVisitor.toInfo(), residualHeapVisitor.generatorDAG, residualHeapVisitor.inspector];
         })();
         if (this.logger.hasErrors()) return undefined;
 
+        let residualOptimizedFunctions = new ResidualOptimizedFunctions(
+          generatorDAG,
+          additionalFunctionValuesAndEffects,
+          residualHeapInfo.values
+        );
+        let referentializer = new Referentializer(
+          this.realm,
+          this.options,
+          preludeGenerator.createNameGenerator("__scope_"),
+          preludeGenerator.createNameGenerator("__get_scope_binding_"),
+          preludeGenerator.createNameGenerator("__leaked_"),
+          residualOptimizedFunctions
+        );
+        for (let instance of residualHeapInfo.functionInstances.values()) referentializer.referentialize(instance);
```

Comment:
This is now like a separate phase. Can you wrap it into something like
```js
statistics.referentialization.measure(() => 
```
similar to `statistics.deepTraversal`? Shouldn't take much time, but we don't want to ignore it.